Fixed #2272: Compilation failure with C++20

codeinred opened the pull request:
## Description
This commit fixes issue #2272: Compilation failure with C++20. 

To do this, it removes using-declarations for 

- `std::allocator<T>::address()`,
- `std::allocator<T>::construct()`, 
- `std::allocator<T>::max_size()`, and 
- `std::allocator<T>::destroy()` 

from the CustomAllocator class in Matchers.tests.cpp. [These functions were deprecated in C++17, and removed in C++20](https://en.cppreference.com/w/cpp/memory/allocator), so having a using declaration for them causes compilation error. The functions weren't used in the Catch2 code base in earlier versions of C++, so it's not a breaking change. 

This commit also wraps the use of a comma operator inside the bracket operator in *Message.tests.cpp*. This change is necessary because gcc was issuing an error when compiling tests with `-std=c++20`. 

```cpp
CAPTURE(std::vector<int>{1, 2, 3}[0, 1, 2],
```
Becomes
```cpp
CAPTURE(std::vector<int>{1, 2, 3}[(0, 1, 2)],
```
<details><summary>Click to see compilation error</summary>

```
projects/CMakeFiles/SelfTest.dir/SelfTest/UsageTests/Message.tests.cpp.o 
/usr/bin/c++  -I/home/alecto/cowork/Catch2/include -std=c++20 -Wall -Wextra -Wunreachable-code -Wpedantic -Wmissing-declarations -Werror -MD -MT projects/CMakeFiles/SelfTest.dir/SelfTest/UsageTests/Message.tests.cpp.o -MF projects/CMakeFiles/SelfTest.dir/SelfTest/UsageTests/Message.tests.cpp.o.d -o projects/CMakeFiles/SelfTest.dir/SelfTest/UsageTests/Message.tests.cpp.o -c /home/alecto/cowork/Catch2/projects/SelfTest/UsageTests/Message.tests.cpp
In file included from /home/alecto/cowork/Catch2/include/catch.hpp:54,
                 from /home/alecto/cowork/Catch2/projects/SelfTest/UsageTests/Message.tests.cpp:9:
/home/alecto/cowork/Catch2/projects/SelfTest/UsageTests/Message.tests.cpp: In function ‘void ____C_A_T_C_H____T_E_S_T____76()’:
/home/alecto/cowork/Catch2/projects/SelfTest/UsageTests/Message.tests.cpp:245:40: error: top-level comma expression in array subscript is deprecated [-Werror=comma-subscript]
  245 |     CAPTURE(std::vector<int>{1, 2, 3}[0, 1, 2],
      |                                        ^
cc1plus: all warnings being treated as errors
```
</details>

This commit also ran `scripts/approve.py`, which updated *compact.sw.approved.txt*, *console.sw.approved.txt*, and *xml.sw.approved.txt* to reflect the change made in *Message.tests.cpp*.

## GitHub Issues

This PR was motivated by #2272

codeinred committed:
Fixed #2272: Compilation failure with C++20

This commit fixes issue #2272: Compilation failure with C++20. To
do this, it removes using-declarations for deprecated functions
std::allocator<T>::address(), construct(), max_size(), and destroy()
from the CustomAllocator class in Matchers.tests.cpp; and it
wraps the use of a comma operator inside the bracket operator in
Message.tests.cpp. This commit also updates compact.sw.approved.txt,
console.sw.approved.txt, and xml.sw.approved.txt to reflect the change
made in Message.tests.cpp.

Signed-off-by: Alecto Irene Perez <perez.cs@pm.me>

codeinred committed:
Amended previous commit

Ah, MSVC, my old nemesis, we meet again. This commit amends the
previous commit to re-introduce the following using declarations inside
the CustomAllocator class:

- using std::allocator<T>::address;
- using std::allocator<T>::construct;
- using std::allocator<T>::max_size;
- using std::allocator<T>::destroy;

When building with MSVC and using a C++ standard older than
C++17. These were re-introduced because MSVC throws an error if they're
not present when compiling with C++11 or C++14.

Signed-off-by: Alecto Irene Perez <perez.cs@pm.me>

horenmar commented on projects/SelfTest/UsageTests/Matchers.tests.cpp:
```diff
@@ -242,12 +242,14 @@ namespace { namespace MatchersTests {
+#if defined(_MSC_VER) && _MSVC_LANG < 201703L
             using std::allocator<T>::address;
-            using std::allocator<T>::allocate;
             using std::allocator<T>::construct;
-            using std::allocator<T>::deallocate;
             using std::allocator<T>::max_size;
             using std::allocator<T>::destroy;
```
The deprecated ones should be safe to just delete afaik.

codeinred replied:
Hi horenmar,

Thank you so much for your review! I initially deleted the deprecated methods, and this worked for GCC and Clang, but deleting those methods broke the build on older versions of MSVC (MSVC 2015). That's why I added a ifdef specifically to check the version of MSVC being used to compile it.

horenmar replied:
Right. Yeah, then the fix is correct.

For v3 I deleted the deprecated usings, but also v3 no longer supports VS 2015.

horenmar commented on projects/SelfTest/UsageTests/Message.tests.cpp:
```diff
@@ -242,7 +242,7 @@ std::ostream& operator<<(std::ostream& out, helper_1436<T1, T2> const& helper) {
 
 TEST_CASE("CAPTURE can deal with complex expressions involving commas", "[messages][capture]") {
-    CAPTURE(std::vector<int>{1, 2, 3}[0, 1, 2],
+    CAPTURE(std::vector<int>{1, 2, 3}[(0, 1, 2)],
```
This should be instead handled by a localized warning suppression instead. The point of the test is to check that `CAPTURE` handles various weird expressions properly, including commas directly in subscript operator.

horenmar reviewed (changes requested):
Thanks, but the fixes should be done a bit differently (see comments)

codeinred committed:
Added localized warning suppression

This commit restores the previous state of 'CAPTURE can deal with
complex expressions involving commas', and instead uses localized
warning suppression to disable messages warnings about the use of a
comma inside a bracket operator.

Signed-off-by: Alecto Irene Perez <perez.cs@pm.me>

codeinred commented:
> This should be instead handled by a localized warning suppression instead. The point of the test is to check that CAPTURE handles various weird expressions properly, including commas directly in subscript operator.

I added localized warning suppression and restored the initial state of the test!

> The deprecated ones should be safe to just delete afaik.

That was my initial assumption too, but this broke in MSVC 2015, which checked for the existence of those methods. Since these methods become deprecated in C++17, I added a `#if` to check for that.

horenmar commented:
Thanks
